### PR TITLE
Track usage

### PR DESCRIPTION
**Pull request by P0lip**

### CODE REVIEW

File: test/specs/usage/object-source.spec.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+const { expect } = require("chai");
+const $RefParser = require("../../..");
+const path = require("../../utils/path");
+const setupHttpMocks = require("../../utils/setup-http-mocks");
+const { get } = require("../../../lib/bundle/util/object.js");
+
+describe("Usage", () => {
+  beforeEach(() => {
+    setupHttpMocks({
```
Comment (pytlesk4) [inline test/specs/usage/object-source.spec.js:11]: hmm, Setupmocks seems like something cool, going to see what it does. This is a note for myself haha.

File: test/specs/usage/object-source.spec.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  });
+
+  it("should track usage of $refs", async () => {
+    let parser = new $RefParser();
+    const schema = await parser.dereference(path.rel("specs/usage/definitions/document.json"));
+
+    expect(get(schema, "#/properties/books/oneOf/0")).to.deep.equal({ title: "Book v1" });
+    expect(get(schema, "#/properties/books/oneOf/1")).to.deep.equal({ title: "Book v1 (mid=1)" });
+    expect(get(schema, "#/properties/books/oneOf/2")).to.deep.equal({ title: "Book v2" });
+
```
Comment (pytlesk4) [inline test/specs/usage/object-source.spec.js:26]:
```suggestion

    expect(schema).toMatchSnapshot();
```

I just want to see what schema looks like, so I am going to generate a snapshot, this is temporary.

**Review by pytlesk4 — COMMENTED** [review]
Exploring

**Review by pytlesk4 — APPROVED** [review]
Approving just in case what I am assuming is correct. Also, if correct this is exactly what I what we needed. 🤞 

Assumption => The key in the propertyMap is the path where the ref was resolved(aka the parentPath that we talked about), and its value is the ref that was resolved.

With this I should be able to simplify the export api, all I need to do now is resolve the node, with one instance of the json schema ref parser, then look over the keys(parentPath) of the propertyMap and if the value of the propertyMap[key] is to our export API and it has a mask id, get the mask, apply it to the resolved node. The only thing we need to change about applying the mask is taking the parentPath into account.

**Comment by P0lip** [comment]
> Assumption => The key in the propertyMap is the path where the ref was resolved(aka the parentPath that we talked about), and its value is the ref that was resolved.

Right. It's a parent path scoped to a resolved object:

```ts
const parser = new $RefParser();
const resolved = parser.dereference(obj);

parsed.propertyMap; // this will contain property paths mapped to `resolved`, and not `obj`.

// resolved is what we care about.
```

**Comment by P0lip** [comment]
There's something off with Karma tests, with mocking to be precise, that makes tests to fail randomly.
I looked into it, but would need to spend a bit of time to figure it out properly.
Will merge it for the time being, so you can use the code.
I'll get back to it at some point.

**Pull request merged**